### PR TITLE
Image Gallery Search UI

### PR DESCRIPTION
**Pull request by edalex-ian**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [x] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds the UI layer for the new Image Gallery. It is hoped that it will mostly make up the UI elements also required to support video (which is the next piece of work).

The UX has not be finalised, as we've decided to first get this part in so that we can then explore how we may wish to improve it. There are also some further user stories around this, such as disabling the MIME type Search Filters when in gallery mode.

I apologise for this size of this. It grew more than expected as things progressed.

https://user-images.githubusercontent.com/43919233/111256139-86239180-866c-11eb-9fdf-aa481a54a872.mp4

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Review by PenghaiZhang — COMMENTED** [review]
Thanks Ian!  The video looks very nice ! I have left some questions.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/GallerySearchResult.tsx [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+   */
+  items: GallerySearchResultItem[];
+}
+
+const useStyles = makeStyles((_: Theme) =>
+  createStyles({
+    titleBar: {
+      background:
+        "linear-gradient(to top, rgba(0,0,0,0.5) 0%, " +
+        "rgba(0,0,0,0.1) 70%, rgba(0,0,0,0) 100%)",
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/GallerySearchResult.tsx:57]: Sure can. Thanks. :+1:

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/GallerySearchResult.tsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    imgSrc: string,
+    altText: string,
+    onClick: () => void
+  ) => (
+    <GridListTile key={key} onClick={onClick} aria-label={ariaLabel}>
+      <img style={{ cursor: "zoom-in" }} src={imgSrc} alt={altText} />
+      <GridListTileBar
+        className={classes.titleBar}
+        actionIcon={
+          <IconButton
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/GallerySearchResult.tsx:89]: Yes, possibly. That wasn't around when I stated this branch. Let me have a look.

[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/GallerySearchResult.tsx:89]: Looks good, and I like the bonus tooltip.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/GallerySearchResult.tsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    items.flatMap(
+      ({
+        additionalEntries,
+        mainEntry,
+        name,
+        uuid,
+        version,
+      }: GallerySearchResultItem) => {
+        const itemName = name ?? uuid;
+        const onClick = ({
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/GallerySearchResult.tsx:117]: Yeah nah, that's fair. :wink: 

Done. :+1:

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPage.tsx [hunk 4]
```diff
@@ -291,6 +318,7 @@ const SearchPage = ({ updateTemplate }: TemplateUpdateProps) => {
                 sortOrder:
                   queryStringSearchOptions.sortOrder ??
                   searchSettings.defaultSearchSort,
+                displayMode: "list",
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPage.tsx:321]: Ah! That is interesting.

The display mode is happily persisted in the history, but yes, haven't looked at sharing it. :disappointed: 

Will organise to do that as another feature in the future.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPage.tsx [hunk 5]
```diff
@@ -318,14 +346,28 @@ const SearchPage = ({ updateTemplate }: TemplateUpdateProps) => {
...
     if (state.status === "searching") {
+      const doSearch = async (
+        options: SearchPageOptions
+      ): Promise<SearchPageSearchResult> => {
+        switch (options.displayMode) {
+          case "gallery-image":
+          case "gallery-video": // Coming soon
+            return {
+              from: "gallery-search",
+              content: await imageGallerySearch(options),
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPage.tsx:357]: Yup, just until we add the video gallery support, so as to still have the selector work on all options I'm just doing the switch fall-through. Will be reworked very shortly as part of the video gallery work.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPage.tsx [hunk 6]
```diff
@@ -352,9 +394,11 @@ const SearchPage = ({ updateTemplate }: TemplateUpdateProps) => {
   // involve legacy content.
   useEffect(() => {
     if (state.status === "success" && isSelectionSessionInStructured()) {
-      state.result.results.forEach(({ uuid }) => {
-        prepareDraggable(uuid);
-      });
+      state.result.content.results.forEach(
+        ({ uuid }: OEQ.Search.SearchResultItem | GallerySearchResultItem) => {
+          prepareDraggable(uuid);
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPage.tsx:399]: Ah yes, this was early in the piece and was just aiming for basic parity. Will remove that me thinks... Somehow...

File: Source/Plugins/Core/com.equella.core/js/__stories__/search/DisplayModeSelector.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  DisplayModeSelectorProps,
+} from "../../tsrc/search/components/DisplayModeSelector";
+
+export default {
+  title: "Search/DisplayModeSelector",
+  component: DisplayModeSelector,
+} as Meta<DisplayModeSelectorProps>;
+
+const commonParams = {
+  onChange: action("onChange"),
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/__stories__/search/DisplayModeSelector.stories.tsx:31]: Hmmm, not sure I've done that before, so I gave it a go.

It all worked nice, but alas then clicking the buttons didn't result in the normal logging into the 'Actions' tab (see below screenshot) but instead triggered errors in the console log saying `onChange` could not be found. So I guess with `ArgTypes` you need to use this differently?

image

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 8]
```diff
@@ -879,3 +885,79 @@ describe("Add and remove favourite Item,", () => {
...
+
+  let page: RenderResult;
+
+  const queryListItems = () => page.queryAllByLabelText(listItemAriaLabel);
+
+  const queryGalleryItems = () =>
+    page.queryAllByLabelText(galleryItemAriaLabel);
+
+  const isChecked = (label: string): boolean => {
+    const button = screen.getByLabelText(label);
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:908]: Ah yes, this is from when tests were giving me grief. Reverted.

Thank you.

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx [hunk 9]
```diff
@@ -879,3 +885,79 @@ describe("Add and remove favourite Item,", () => {
...
+    async (mode: string) => {
+      expect(queryListItems().length).toBeGreaterThan(0);
+      expect(queryGalleryItems()).toHaveLength(0);
+
+      // Monitor the search function, and change the mode
+      const gallerySearchPromise = mockGallerySearch.mockResolvedValue(
+        transformedBasicImageSearchResponse
+      );
+      await changeMode(mode);
+      await gallerySearchPromise;
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:953]: On my machine, there appears to be no issue. **But** is that just an async race condition waiting to fail and make the test flakey? Or are you saying the `act()` in `changeMode()` is sufficient to protect from that?

**Comment by edalex-ian** [comment]
Thanks for the feedback @PenghaiZhang . All feedback other then the two still marked as not resolved have been addressed (in comment or code).

Please have a look and let me know what you think. Also, interested to hear your responses to the two remaining items.

File: Source/Plugins/Core/com.equella.core/js/__stories__/search/DisplayModeSelector.stories.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__stories__/search/DisplayModeSelector.stories.tsx:31]: Hmmm, I am not quite sure why you have this error. Can you try again on `CategorySelector` ? 
image

File: Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/js/__tests__/tsrc/search/SearchPage.test.tsx:953]: Yes, I have a feeling that the act() in changeMode() is sufficient, but I don't mind to wait for the promise.

File: Source/Plugins/Core/com.equella.core/js/__stories__/search/DisplayModeSelector.stories.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/__stories__/search/DisplayModeSelector.stories.tsx:31]: Yeah, that one works fine for me. I'll have more of a look tomorrow to see how they differ. :thinking:

[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/__stories__/search/DisplayModeSelector.stories.tsx:31]: Aha! I missed that the syntax is slightly different. That is, `action` is now a property and not a function.

**Comment by edalex-ian** [comment]
Okay @PenghaiZhang . All done. Let me know if you're good, and if so on green I'll merge so people can start having a play with this.

**Review by PenghaiZhang — APPROVED** [review]
LGTM!!

**Pull request merged**